Enhancement: Allow re-use of router instances

### Summary

Mounting a router to multiple apps raises:

```
E               litestar.exceptions.http_exceptions.ImproperlyConfiguredException: 500: Router with path / has already been registered
```

There's been some discussion at least in https://github.com/litestar-org/litestar/issues/148 and https://github.com/litestar-org/litestar/issues/1054 and apparently it's not a bug, but intentional. But why does the limitation exist?

The problem is trying to re-use the same router instance when re-creating the application in tests. The first test passes, as the app constructs okay, but when the code is trying to re-build the application for the second test, the error is thrown. This totally kills the ability to create applications with altered configurations in tests.

This is only for the `Router` instances. If you map controllers directly, then there's no issue.

Could the limitation be lifted?

### Basic Example

```py
class MyHandler(Controller): ...

router = Router(route_handlers=[MyHandler])

def create_app():
    return Litestar(route_handlers=[router])

@pytest.fixture(scope="function")
def client():
    return TestClient(app=create_app())

def test_one(client):
    pass

def test_two(client):
    pass
```

### Drawbacks and Impact

_No response_

### Unresolved questions

_No response_

<!-- POLAR PLEDGE BADGE START -->
---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3012">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3012/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/3012/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

--- a/litestar/router.py
+++ b/litestar/router.py
@@ -1,7 +1,7 @@
 from __future__ import annotations
 
 from collections import defaultdict
-from copy import copy
+from copy import copy, deepcopy
 from typing import TYPE_CHECKING, Any, Mapping, Sequence, cast
 
 from litestar._layers.utils import narrow_response_cookies, narrow_response_headers
@@ -309,14 +309,12 @@ class Router:
             return value(owner=self).to_handler()  # pyright: ignore
 
         if isinstance(value, Router):
-            if value.owner:
-                raise ImproperlyConfiguredException(f"Router with path {value.path} has already been registered")
-
             if value is self:
                 raise ImproperlyConfiguredException("Cannot register a router on itself")
 
-            value.owner = self
-            return value
+            router_copy = deepcopy(value)
+            router_copy.owner = self
+            return router_copy
 
         if isinstance(value, (ASGIRouteHandler, HTTPRouteHandler, WebsocketRouteHandler)):
             value.owner = self

--- a/tests/e2e/test_router_registration.py
+++ b/tests/e2e/test_router_registration.py
@@ -132,9 +132,7 @@ def test_register_validation_wrong_class() -> None:
 def test_register_already_registered_router() -> None:
     first_router = Router(path="/first", route_handlers=[])
     Router(path="/second", route_handlers=[first_router])
-
-    with pytest.raises(ImproperlyConfiguredException):
-        Router(path="/third", route_handlers=[first_router])
+    Router(path="/third", route_handlers=[first_router])
 
 
 def test_register_router_on_itself() -> None:

